libs/language-server/src/lib/ast/expressions/operators/binary-arithmetic-expression.ts

```diff
@@ -0,0 +1,202 @@
+// SPDX-FileCopyrightText: 2023 Friedrich-Alexander-Universitat Erlangen-Nurnberg
+//
+// SPDX-License-Identifier: AGPL-3.0-only
+
+import { strict as assert } from 'assert';
+
+import { ValidationContext } from '../../../validation/validation-context';
+import { BinaryExpression } from '../../generated/ast';
+// eslint-disable-next-line import/no-cycle
+import {
+  PropertyValuetype,
+  isNumericType,
+  numericTypes,
+} from '../../model-util';
+import { evaluateExpression } from '../evaluation';
+import {
+  BinaryTypeInferenceFunction,
+  EvaluationFunction,
+  EvaluationStrategy,
+} from '../operator-registry';
+import { generateUnexpectedTypeMessage } from '../type-inference';
+
+export const inferBinaryArithmeticExpressionType: BinaryTypeInferenceFunction =
+  (
+    leftType: PropertyValuetype,
+    rightType: PropertyValuetype,
+    expression: BinaryExpression,
+    context: ValidationContext | undefined,
+  ): PropertyValuetype | undefined => {
+    assert(
+      expression.operator === '+' ||
+        expression.operator === '-' ||
+        expression.operator === '*' ||
+        expression.operator === '/' ||
+        expression.operator === '%',
+    );
+
+    if (!isNumericType(leftType) || !isNumericType(rightType)) {
+      if (!isNumericType(leftType)) {
+        context?.accept(
+          'error',
+          generateUnexpectedTypeMessage(numericTypes, leftType),
+          {
+            node: expression.left,
+          },
+        );
+      }
+      if (!isNumericType(rightType)) {
+        context?.accept(
+          'error',
+          generateUnexpectedTypeMessage(numericTypes, rightType),
+          {
+            node: expression.right,
+          },
+        );
+      }
```

Comment:
Could be extracted to a separate method to improve the reading flow of the method. Personally, I think nested control structures are somewhat hard to read.